File: 0308-Range Sum Query 2D - Mutable/0308-Range Sum Query 2D - Mutable.cpp
class NumMatrix {
public:
    NumMatrix(vector<vector<int>> matrix) {
        int m = matrix.size();
        if (m == 0) {
            return;
        }
        
        int n = matrix[0].size();
        bit = vector<vector<int>>(1 + m, vector<int>(1 + n));
        matrixCopy = vector<vector<int>>(m, vector<int>(n));
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                update(i, j, matrix[i][j]);
            }
        }
    }
    
    void update(int row, int col, int val) {
        int diff = val - matrixCopy[row][col];
        matrixCopy[row][col] = val;
        for (int i = row + 1; i <= matrixCopy.size(); i += (-i) & i) {
            for (int j = col + 1; j <= matrixCopy[i - 1].size(); j += (-j) & j) {
                bit[i][j] += diff;
            }
        }
    }
    
    int sumRegion(int row1, int col1, int row2, int col2) {
        return read(row2 + 1, col2 + 1) - read(row1, col2 + 1) - read(row2 + 1, col1) + read(row1, col1);
    }
    
private:
    vector<vector<int>> bit;
    vector<vector<int>> matrixCopy;
    int read(int x, int y) {
        int sum = 0;
        for (int i = x; i > 0; i -= (-i) & i) {
            for (int j = y; j > 0; j -= (-j) & j) {
                sum += bit[i][j];
            }
        }
        return sum;
    }
};

/**
 * Your NumMatrix object will be instantiated and called as such:
 * NumMatrix obj = new NumMatrix(matrix);
 * obj.update(row,col,val);
 * int param_2 = obj.sumRegion(row1,col1,row2,col2);
 */
 

File: 0301-0400/0308-Range Sum Query 2D - Mutable/0308-Range Sum Query 2D - Mutable.py
class NumMatrix:

    def __init__(self, matrix: List[List[int]]):
        if not matrix:
            return
        m, n = len(matrix), len(matrix[0])
        self.mat = [[0] * n for _ in range(m)]
        self.bit = [[0] * (1 + n) for _ in range(1 + m)]
        for i in range(m):
            for j in range(n):
                self.update(i, j, matrix[i][j])

    def update(self, row: int, col: int, val: int) -> None:
        diff = val - self.mat[row][col]
        if diff != 0:
            self.mat[row][col] = val
            i = row + 1
            while i < len(self.bit):
                j = col + 1
                while j < len(self.bit[0]):
                    self.bit[i][j] += diff
                    j += (-j) & j
                i += (-i) & i

    def sumRegion(self, row1: int, col1: int, row2: int, col2: int) -> int:
        return self.read(row2 + 1, col2 + 1) - self.read(row1, col2 + 1) - self.read(row2 + 1, col1) + self.read(row1, col1)

        
    def read(self, r, c):
        total = 0
        i = r
        while i > 0:
            j = c
            while j > 0:
                total += self.bit[i][j]
                j -= (-j) & j
            i -= (-i) & i
        return total
        

# Your NumMatrix object will be instantiated and called as such:
# obj = NumMatrix(matrix)
# obj.update(row,col,val)
# param_2 = obj.sumRegion(row1,col1,row2,col2)
